application.go

```diff
@@ -57,3 +64,120 @@ func (ac *ApplicationClient) GetApplication(
+		appRing := ApplicationRing{Application: application}
+		// Get the gateways that are delegated to the application
+		// at the query height and check if the given gateway address is in the list.
+		gatewaysDelegatedTo := appRing.ringAddressesAtBlock(queryHeight)
+		if slices.Contains(gatewaysDelegatedTo, gatewayAddress) {
+			// The application is delegating to the given gateway address, add it to the list.
+			gatewayDelegatingApplications = append(gatewayDelegatingApplications, application.Address)
+		}
+	}
+
+	return gatewayDelegatingApplications, nil
+}
+
+type ApplicationRing struct {
+	types.Application
+	PublicKeyFetcher
+}
+
+// GetRing returns the ring for the application.
+// The ring is created using the application's public key and the public keys of
+// the gateways that are currently delegated from the application.
+func (a ApplicationRing) GetRing(
+	ctx context.Context,
+	queryHeight uint64,
+) (addressRing *ring.Ring, err error) {
+	// Get the gateway addresses that are delegated from the application
+	// at the query height.
+	currentGatewayAddresses := a.ringAddressesAtBlock(queryHeight)
+
+	ringAddresses := make([]string, 0)
+	ringAddresses = append(ringAddresses, a.Application.Address)
+
+	// If there are no current gateway addresses, use the application address as the ring address.
+	if len(currentGatewayAddresses) == 0 {
+		ringAddresses = append(ringAddresses, a.Application.Address)
+	} else {
+		ringAddresses = append(ringAddresses, currentGatewayAddresses...)
+	}
+
+	curve := ring_secp256k1.NewCurve()
+	ringPoints := make([]ringtypes.Point, 0, len(ringAddresses))
+
+	// Create a ring point for each address.
+	for _, address := range ringAddresses {
+		pubKey, err := a.PublicKeyFetcher.GetPubKeyFromAddress(ctx, address)
+		if err != nil {
+			return nil, err
+		}
+
+		point, err := curve.DecodeToPoint(pubKey.Bytes())
+		if err != nil {
+			return nil, err
+		}
+
+		ringPoints = append(ringPoints, point)
+	}
+
+	return ring.NewFixedKeyRingFromPublicKeys(ring_secp256k1.NewCurve(), ringPoints)
+}
+
+func (a ApplicationRing) ringAddressesAtBlock(
+	queryHeight uint64,
+) []string {
+	// Get the current active delegations for the application and use them as a base.
+	activeDelegationsAtHeight := a.Application.DelegateeGatewayAddresses
+
+	// Use a map to keep track of the gateways addresses that have been added to
+	// the active delegations slice to avoid duplicates.
+	addedDelegations := make(map[string]struct{})
+
+	// Iterate over the pending undelegations recorded at their respective block
+	// height and check whether to add them back as active delegations.
+	for pendingUndelegationHeight, undelegatedGateways := range a.Application.PendingUndelegations {
+		// If the pending undelegation happened BEFORE the target session end height, skip it.
+		// The gateway is pending undelegation and simply has not been pruned yet.
+		// It will be pruned in the near future.
+		if pendingUndelegationHeight < queryHeight {
```

Comment:
This check should be done against the session's end height that we derive from the provided block height.

Any reason why we are re-implementing this while we already have an implementation on `poktroll`?

I believe that this will add more maintenance burden if things change on the protocol side.

Reply:
No arguments against using Poktroll to get the application's ring, except the current Poktroll implementation requires `sharedtypes.Params` as input, which is only used to get the session ending height.

The implementation in this PR scraps that requirement and uses the passed queryHeight, which the caller can get from the session. Having a similar function on Poktroll is of course preferable.

To potentially take this a step further, considering a session has a specific ending height and is calculated for a specific application, shouldn't the application's ring also be a method offered by the session?